### DISCUSSION TITLE
H96MaxW2 box (S905W2 Experimental)

### DISCUSSION THREAD

**Discussion by markosc** [devmfc/debian-on-amlogic#20] [Q&A]
Hello, I have an s905w2 TV box H96MAX 4g / 64gb . I can't get it to boot with the generic image, but I can boot it with CoreELEC using the variant s4_s905w2_4g, and booting with USB instead of microSD.
![image](https://github.com/devmfc/debian-on-amlogic/assets/16872280/5da2dc30-cc69-4506-a5db-36562602581c)

**Comment by markosc** [comment]
boot, no HDMI image display, and it doesn't detect WiFi and USB 3.0, only USB 2.0,

**Comment by LITUATUI** [comment]
Hi. I have the same box, have you managed to get HDMI working?

**Comment by devmfc** [comment]
It looks like these boxes are a bit different than the first generation with S905W2 SOC's. 

They will boot, but indeed, no HDMI output, even not with the hack I use for the other S905W2 boxes (different memory address I guess). 

WIFI is a real no go.  My H96Max W2 box (2G) has some WIFI module which the VENDEV ID points to RDA Microelectronics RDA5991, I never heard of such wifi module and google comes up empty. But maybe the database I use is not correct..  eMMC does work, so does USB2 (both ports) and Ethernet.

When I have time in a few weeks I will take a look at console output and will try to make it work. 

But don't expect much from this box... They are really cheap for a reason.

**Reply by LITUATUI** [reply]
I really like it as it costed me only 20 euros on AliExpress for the 4/64 GB version and it has WiFi 6. It also runs Android TV really well with [slimBOXtv firmware](https://slimboxtv.ru/h96-max-w2/). At this price there's nothing better.

Last week I ordered a Transpeed X4 (S905X4 SoC) to play with, maybe I'll get better luck having stable Debian on it.

Thank you for porting Debian to these cheap devices.

**Reply by devmfc** [reply]
If you just like to tinker with these boxes, they are just fine of course.

You know that S905X4 is also very Experimental? 
If you want stable Debian I'd try an S905X3 based box.

Have fun!

**Reply by LITUATUI** [reply]
Thanks mate, I know that S905X3 would provide better compatibility, but nowadays I don't buy devices without AV1 hardware decoding.

If I can't make Debian and CasaOS run stable on it, I'll just use CoreELEC for the docker containers. CoreELEC runs great on S905X4.

**Reply by markosc** [reply]
As LITUATUI mentions, I also have that box priced at 20 dollars, with 4GB/64GB of space. While it's interesting to use it with Debian and set it up as a server, what surprises me is that it supposedly has a USB 3.0 port and is colored blue, very similar to the other version, H96 Max (RK3318). Well, upon booting up the system, it only detects a USB 2.0 port and not the 3.0

**Comment by denisidoro** [comment]
Not sure if it helps, but I'm also able to boot CoreELEC, even though there's no wifi there either

**Reply by devmfc** [reply]
The goal of my images is not o be a media image. In fact the goal is to have Debian/Ubuntu images with a mainline LTS kernel.  CoreElec is a completely different distribution and based on the old (and buggy) vendor kernel. And of course will that kernel support Amlogic hardware better (for now).

The next release of my image will have experimental support for the H96MaxW2 box

**Reply by denisidoro** [reply]
Perfect! Thanks

**Comment by denisidoro** [comment]
@devmfc, I've just tried the latest debian build (https://github.com/devmfc/debian-on-amlogic/releases/tag/v6.6.16) and I'm getting a a kernel panic when trying to boot it from a microSD

![IMG_20240212_214933](https://github.com/devmfc/debian-on-amlogic/assets/3226564/789aa388-c943-4745-b667-1d8e6d768c58)

**Reply by denisidoro** [reply]
I tried booting it from a USB stick and it worked

**Reply by denisidoro** [reply]
I was even able to install Home Assistant supervised on it without any issues 👍 
![shot 2024-02-13 at 08 26 48](https://github.com/devmfc/debian-on-amlogic/assets/3226564/bc301957-ebdf-42db-a3a3-8ab86ec5ee15)

**Comment by Raviu56** [comment]
I can confirm than booting h96max s905w2 4g boot from USB
But `aml-multiboot-setup.sh` does not work
Wifi does not work, I've used a USB wifi.
I did not try hdmi , I'm using it as headless server.

**Reply by devmfc** [reply]
That's goo to hear.
Did you use the same USB drive on another box before booting this one? If so try `source /etc/boot-run-once/03-set_fw_env_config` and try again. If that does not work can you please paste the output of `~/aml-part.sh` here?

**Reply by Raviu56** [reply]
If I try to source it, it instantly exits the shell.
Here is what I think could be useful
```
tvbox:~# bash -x /etc/boot-run-once/03-set_fw_env_config 
++ grep -oP '(?<=\()\w+(?=\))' /sys/bus/soc/devices/soc0/soc_id
+ SOC_ID=S905W2
+ grep -l MMC /sys/bus/mmc/devices/mmc2:4492/type
+ exit
tvbox:~# ls -l /sys/bus/mmc/devices/mmc2:4492/type
-r--r--r-- 1 root root 4096 Feb 26 05:44 /sys/bus/mmc/devices/mmc2:4492/type
tvbox:~# cat /sys/bus/mmc/devices/mmc2:4492/type
SDIO
tvbox:~# ~/aml-part.sh 
od: /dev/mmcblk1: No such file or directory
/root/aml-part.sh: line 76: table[0]: unbound variable
tvbox:~# 
```

I have multiple pieces of this tvbox, I can help troubleshooting it further
Attached the full dmesg
[h96maxS905W2.dmesg.txt](https://github.com/devmfc/debian-on-amlogic/files/14399811/h96maxS905W2.dmesg.txt)

**Reply by Raviu56** [reply]
I've to `adb connect TheIP` , then `adb shell reboot update` to boot it into debian

**Reply by devmfc** [reply]
~~~
mmc1: tuning execution failed: -5
[    0.374451] mmc1: error -5 whilst initialising MMC card
[    0.375762] mmc1: Failed to initialize a non-removable card
~~~
Your eMMC did not initialize properly. That is the reason the scripts fail.

**Reply by Raviu56** [reply]
So how to solve it? I am ready to help solving this issue and the internal wifi issue not working too.
Hoping the dmesg helps, if further debugging is needed I'm ready.

**Reply by devmfc** [reply]
Did you flash your box with some other image before? 
only thing I can think of is ty setting your max-frequency property for your mmc controller to a lower number...

Wifi is a no go, I could not find a source for the wifi driver anywhere for any kernel. Let alone a version for the mainline kernel.

**Reply by Adney** [reply]
I found information about the driver for wifi/bt.They use L8800D-SHCY module, which is based on AIC8800.
It will be good if you add it to your builds

https://discourse.coreelec.org/t/looking-for-users-with-no-working-wifi-or-bt/9649/1041

**Reply by devmfc** [reply]
That looks promising... Thanks for the link!

But CoreElec is not mainline linux. Does it work for mainline Linux?
It is a DKMS package, so you probably need the kernel headers to install the package.

The needed kernel headers are in he release files: https://github.com/devmfc/debian-on-amlogic/releases/download/v6.6.16/linux-headers-6.6.16-meson64_20240212_arm64.deb. You can install them with `apt-get install ./linux-headers-6.6.16-meson64_20240212_arm64.deb`.

If it works let me know, then I can try to add them to the newer images.

**Reply by ductm208** [reply]
If it helps any one:
I have Tanix W2 which S905W2 soldered:
- Toshiba 64GB sd card did not start, got stuck at mounting partitions as same above.
- flash drive 4gb made it
I spent some hrs to see what happened. I want to boot from SDcard to save 2 USB ports. After few attempts with different images and wondered: did it go wrong with the restorations or sdcard was damage? Then I tried 4GB sdcard no name (JVJ labeled but very cheap) surprisingly it made system booted smoothly. 

I guess the issue is from the storage sizes. Anyone tried wih 8, 18, 32 recently?

**Reply by rogerpueyo** [reply]
Hi,

I successfully booted [v6.6.16](https://github.com/devmfc/debian-on-amlogic/releases/tag/v6.6.16) (from a µSD card) and [v6.6.23](https://github.com/devmfc/debian-on-amlogic/releases/tag/v6.6.23) (from a USB flash drive) on the H96 Max W2 box. @devmfc, thanks for building these images and adding support for these devices. Really cool, very much appreciated.

First thing I did was a backup of the whole eMMC to an external storage, which seemed to work (same checksums). Then, I tried installing Debian to eMMC. With both v6.6.16 and v6.6.23, enabling multiboot setup with `/aml-multiboot-setup.sh` fails:
```
tvbox:~#./aml-multiboot-setup.sh 
Setup multiboot. This will reconfigure your bootloader environment to use a 'boot order' to load boot scripts while booting. The default boot order will be set to: 
1. sd, 
2. usb, 
3. emmc, 
4. original boot.

This will change your bootloader environment. By doing so there is always a small risk to 'brick' your box. If you don't have the tools, image and knowledge to fix that, you should reconsider. Personally I have never 'bricked' a box with this script. However ymmv, don't blame me.
Are you sure to proceed? y/Ny
Ok, proceeding
Trying to config u-boot environment storage location... [Ok]
Backupping current u-boot environment to uEnv.bak... [Ok]
Writing boot commands to bootloader... 
Cannot read environment, using default
Cannot read default environment from file
[Error]
fw_setenv
tvbox:~#./aml-multiboot-setup.sh 
Setup multiboot. This will reconfigure your bootloader environment to use a 'boot order' to load boot scripts while booting. The default boot order will be set to: 
1. sd, 
2. usb, 
3. emmc, 
4. original boot.

This will change your bootloader environment. By doing so there is always a small risk to 'brick' your box. If you don't have the tools, image and knowledge to fix that, you should reconsider. Personally I have never 'bricked' a box with this script. However ymmv, don't blame me.
Are you sure to proceed? y/Ny
Ok, proceeding
Trying to config u-boot environment storage location... Could not find a valid u-boot environment storage location.

[Error]
Could not find a valid u-boot environment storage location.
This means that modification of bootloader environment is not possible.
In more than 80% of the cases this is caused by the presence of non-stock u-boot bootloader. Did you flash some other firmware to eMMC?
Please install stock Android firmware and try again.
```

`dmesg` shows:
```
[...]
[    8.192528] meson8b-dwmac fdc00000.ethernet eth0: Link is Up - 100Mbps/Full - flow control rx/tx
[44652.705144] I/O error, dev mmcblk1, sector 237568 op 0x1:(WRITE) flags 0x800 phys_seg 16 prio class 2
[44652.705708] Buffer I/O error on dev mmcblk1, logical block 29696, lost async page write
[44652.706696] Buffer I/O error on dev mmcblk1, logical block 29697, lost async page write
[44652.707698] Buffer I/O error on dev mmcblk1, logical block 29698, lost async page write
[44652.708681] Buffer I/O error on dev mmcblk1, logical block 29699, lost async page write
[44652.709673] Buffer I/O error on dev mmcblk1, logical block 29700, lost async page write
[44652.710663] Buffer I/O error on dev mmcblk1, logical block 29701, lost async page write
[44652.711687] Buffer I/O error on dev mmcblk1, logical block 29702, lost async page write
[44652.712658] Buffer I/O error on dev mmcblk1, logical block 29703, lost async page write
[44652.713652] Buffer I/O error on dev mmcblk1, logical block 29704, lost async page write
[44652.714648] Buffer I/O error on dev mmcblk1, logical block 29705, lost async page write
```

Based on your comments, I imagined the `max-frequency` value should be lowered. I decompiled the `meson-s4-h96maxw2.dtb` DTB and changed the `max-frequency` values of the three `mmc` nodes to `0x989680`(10 MHz). Possibly, only changing the one node for the eMMC would be enough. Anyway. After compiling the new DTB and rebooting, I got the script working (no eMMC errors seen on `dmesg`):

```
tvbox:~#./aml-multiboot-setup.sh 
Setup multiboot. This will reconfigure your bootloader environment to use a 'boot order' to load boot scripts while booting. The default boot order will be set to: 
1. sd, 
2. usb, 
3. emmc, 
4. original boot.

This will change your bootloader environment. By doing so there is always a small risk to 'brick' your box. If you don't have the tools, image and knowledge to fix that, you should reconsider. Personally I have never 'bricked' a box with this script. However ymmv, don't blame me.
Are you sure to proceed? y/Ny
Ok, proceeding
Trying to config u-boot environment storage location... [Ok]
Backupping current u-boot environment to uEnv.bak... [Ok]
Writing boot commands to bootloader... 
Done.
```

Nice!

(edit, continued)

I could even install to eMMC, it seems:
```
tvbox:~#./aml-install-to-emmc.sh gpt
#
#	WARNING: This script CAN (soft) brick your TVBOX and is UNSUPPORTED!
#
#	This script has not been tested extensively, is not high quality and makes a
#	lot of assumptions about your bootloader. If your bootloader version fails
#	to meet any of these assumptions... It will no longer boot from eMMC.
#	However, in most cases, the box will still be able to boot from a USB flash
#	drive or USB otg via the Amlogic USB burning tool. You can often use this
#	method to fix boot problems.
#
#	If you don't have the knowledge or tools to unbrick an Amlogic system and
#	you are very attached to your box, maybe this script isn't for you.
#
#	That being said, it works on my full range of boxes, S905, S905X, S905W, 
#	S912, S905X2, S922X, S905X3, S905X4 ans S905W2. I have never bricked one
#	of my boxes with this script.
#	But of course, every TVBOX can be different, so your mileage may vary.
#
#	Anyway, if you use this script, don't blame me or ask me to fix it 
#	when it goes wrong, you have been warned, you're on your own (and Google).
#

Current Host disk:  [ /dev/sda ]
Target eMMC disk: [ /dev/mmcblk1 ]
Select partition table type [ GPT ]
This script will erase all existing partitions and its data on the disk!!
Are you sure y/N ? y
Ok, proceeding...
Start create partition table and filesystems on eMMC [ /dev/mmcblk1 ]... 
Wipe bootloader from mmc userpart
Partition [env] not found
Waiting for sync... [ ok ]
Model: MMC DF4064 (sd/mmc)
Disk /dev/mmcblk1: 62.5GB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags: 

Number  Start   End     Size    File system  Name      Flags
 3      4194kB  1611MB  1606MB               reserved
 1      1611MB  2147MB  537MB   fat16        boot      msftdata
 2      2147MB  62.5GB  60.4GB  ext4         root

Installing missing package [ dosfstools ]
Selecting previously unselected package dosfstools.
(Reading database ... 16880 files and directories currently installed.)
Preparing to unpack .../dosfstools_4.2-1_arm64.deb ...
Unpacking dosfstools (4.2-1) ...
Setting up dosfstools (4.2-1) ...
mkfs.fat 4.2 (2021-01-31)
mke2fs 1.47.0 (5-Feb-2023)
/dev/mmcblk1p2 contains a ext4 file system labelled 'ROOTFS'
	last mounted on / on Thu Jan  1 01:00:01 1970
Proceed anyway? (y,N) y
Discarding device blocks: done                            
Creating filesystem with 14743296 4k blocks and 3686400 inodes
Filesystem UUID: 2ba22159-d6c0-4039-8eed-a93a08db8a10
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208, 
	4096000, 7962624, 11239424

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (65536 blocks): done
Writing superblocks and filesystem accounting information: done   

[ done ]
 Mounting newly created filesystems... [ done ]
Copy /boot to boot partition... [ done ]
Creating rootfs... [ done ]
Copying rootfs files... 
Installing missing package [ rsync ]
Selecting previously unselected package rsync.
(Reading database ... 16915 files and directories currently installed.)
Preparing to unpack .../rsync_3.2.7-1_arm64.deb ...
Unpacking rsync (3.2.7-1) ...
Setting up rsync (3.2.7-1) ...
rsync.service is a disabled or a static unit, not starting it.
Copying [ /srv ]
              0 100%    0.00kB/s    0:00:00 (xfr#0, to-chk=0/1)
Copying [ /opt ]
              0 100%    0.00kB/s    0:00:00 (xfr#0, to-chk=0/1)
Copying [ /var ]
    232,797,880  99%   33.94MB/s    0:00:06 (xfr#1424, to-chk=0/1498)  
Copying [ /etc ]
      1,260,880  98%    3.73MB/s    0:00:00 (xfr#270, to-chk=0/905)  
Copying [ /sbin ]
              0   0%    0.00kB/s    0:00:00 (xfr#0, to-chk=0/1)
Copying [ /home ]
          4,568 100%    1.45MB/s    0:00:00 (xfr#4, to-chk=0/6)
Copying [ /bin ]
              0   0%    0.00kB/s    0:00:00 (xfr#0, to-chk=0/1)
Copying [ /usr ]
    529,624,310  99%    9.15MB/s    0:00:55 (xfr#13568, to-chk=0/16339)   
Copying [ /lib ]
              0   0%    0.00kB/s    0:00:00 (xfr#0, to-chk=0/1)
Copying [ /boot ]
     36,864,702 100%   71.69MB/s    0:00:00 (xfr#293, to-chk=0/298)  
Copying [ /root ]
         50,587 100%    5.88MB/s    0:00:00 (xfr#12, to-chk=0/14)
[ done ]
Creating new fstab... [ done ]
Updating boot.config... [ done ]
Wait for sync... [ done ]
Cleaning up...  [done] 

Ready, you can now reboot. Or not, whatever you want.
```

**Reply by denisidoro** [reply]
@devmfc any chance you could change the `max-frequency` so that this works out-of-the-box in the next release? Thanks!

**Reply by devmfc** [reply]
@rogerpueyo You're welcome and thanks for reporting and troubleshooting this.

And yes, I will lower the max-frequency, but 10Mhz is really really slow.  
Question is, what are reasonable values for these boxes...

Maybe someone can try the three attached different dtb's with different frequencies?


[meson-s4-h96maxw2-dtb.zip](https://github.com/devmfc/debian-on-amlogic/files/14972036/meson-s4-h96maxw2-dtb.zip)

**Reply by rogerpueyo** [reply]
@devmfc,

I gave your three DTBs a shot. Unfortunately, none worked; they all gave the same Buffer I/O error:
```
meson-s4-h96maxw2-50-35-100.dtb
meson-s4-h96maxw2-100-35-100.dtb
meson-s4-h96maxw2-100-35-167.dtb
Ok, proceeding
Trying to config u-boot environment storage location... [Ok]
Backupping current u-boot environment to uEnv.bak... [Ok]
Writing boot commands to bootloader... 
[  140.542563] I/O error, dev mmcblk1, sector 237568 op 0x1:(WRITE) flags 0x800 phys_seg 16 prio class 2
[  140.543138] Buffer I/O error on dev mmcblk1, logical block 29696, lost async page write
[  140.544108] Buffer I/O error on dev mmcblk1, logical block 29697, lost async page write
[  140.545087] Buffer I/O error on dev mmcblk1, logical block 29698, lost async page write
[  140.546083] Buffer I/O error on dev mmcblk1, logical block 29699, lost async page write
[  140.547102] Buffer I/O error on dev mmcblk1, logical block 29700, lost async page write
[  140.548085] Buffer I/O error on dev mmcblk1, logical block 29701, lost async page write
[  140.549074] Buffer I/O error on dev mmcblk1, logical block 29702, lost async page write
[  140.550070] Buffer I/O error on dev mmcblk1, logical block 29703, lost async page write
[  140.551093] Buffer I/O error on dev mmcblk1, logical block 29704, lost async page write
[  140.552064] Buffer I/O error on dev mmcblk1, logical block 29705, lost async page write
[  140.560365]  mmcblk1: p1 p2 p3
Cannot read environment, using default
Cannot read default environment from file
[Error]
```

Would you please share the DTS file, so that I can generate DTBs to find which is the maximum supported frequency?

**Reply by rogerpueyo** [reply]
Hi again, @devmfc,

In my first message, a few days ago, I was setting `max-frequency` to 10 MHz and the eMMC just worked fine. I tested several values for the `max-frequency` setting of the `mmc@fe08c000` node, with the following results:
 - 10/20 MHz: OK
 - 25/30/40/50/100 MHz: I/O error

The test I ran was very simple. First I made a backup of /dev/mmcblk1p1 partition to a file on a USB drive using `dd`. Afterwards, I tried to write it back to the eMMC for each of the frequencies above. As you can see, anything beyond 20 MHz failed.

**Reply by devmfc** [reply]
@rogerpueyo To get the source, (dts) simply run `dtc xyz.dtb > xyz.dts`. If dtc is not present `apt-get install device-tree-compiler`

20 Mhz is reallyyyyyy slooooowww.  I'm afraid the mainline driver is not suitable for the newer generation mmc controllers, or I'm missing something in my DTB's. And I'm not going to port the current vendor driver to mainline, these drivers are really.... uh, let's say complicated.

Attached a version with 20Mhz for eMMC. But I'm not going to put that is my images as the default DTB, especially while my H96MaxW2 box AFAICR does work with higher frequencies... I'll have to try that again to confirm that somewhere in the coming weeks. 

[meson-s4-h96maxw2-50-35-20.zip](https://github.com/devmfc/debian-on-amlogic/files/15016957/meson-s4-h96maxw2-50-35-20.zip)

**Reply by rogerpueyo** [reply]
Thanks, @devmfc ,

`meson-s4-h96maxw2-50-35-20.dtb` works correctly.

> To get the source, (dts) simply run dtc xyz.dtb > xyz.dts.

Great, I did it like this, but I got a few warnings. Hence, I was wondering if I was missing anything.

> 20 Mhz is reallyyyyyy slooooowww. 
```
tvbox:~#dd if=/root/dev-mmcblk1p1.dd of=/dev/mmcblk1p1 status=progress
535015936 bytes (535 MB, 510 MiB) copied, 120 s, 4.5 MB/s
```
This is the write speed I'm getting at 20 MHz. The source file, on a USB stick, is read at max. 20 MB/s (that shouldn't be the bottleneck).

Please let me know if I can help somehow.

**Reply by denisidoro** [reply]
Hey, folks! Just wanted to thank you for making progress on this. I'm afraid I'm unable to help, but wanted to let you know that more people will benefit from this.

**Reply by denisidoro** [reply]
@devmfc, I'm afraid you forgot to include the `-slow.dtb` in the latest released image. I was unable to boot when uncommenting the `h96max_slow` line. After some troubleshooting, I copied the dtb linked here into the correct directory and it worked

**Reply by devmfc** [reply]
Yeah, I know. Sorry. Thanks for letting me know. They will be included in next release.

**Comment by markosc** [comment]
Can someone confirm if the new update detects USB 3.0?

**Reply by LITUATUI** [reply]
The port is USB 2.0, don't let the blue color fool you.

**Comment by DarKOrange75** [comment]
Hello,
I've flashed a usb stick, chose 
I successfully booted on a USB stick but now trying to install on eMMC fails:

Current Host disk:  [ /dev/sda ]
Target eMMC disk: [ /dev/mmcblk1 ]
Select partition table type [ MSDOS ]
This script will erase all existing partitions and its data on the disk!!
Are you sure y/N ? y
Ok, proceeding...
Start create partition table and filesystems on eMMC [ /dev/mmcblk1 ]... 
Backup vendor u-boot
Found: 27
offset: 742391808
size: 2516582400
dev		nr	name	offset		size	end		flags
Start restoring vendor u-boot
dd: fsync failed for '/dev/mmcblk1': Input/output error
Error occured [1], create_partition_table:127, exitting

dmesg:
   98.656679]  mmcblk1: p1 p2
[   99.018662]  mmcblk1:
[   99.206787]  mmcblk1: p1
[   99.446115]  mmcblk1: p1 p2
[   99.456820]  mmcblk1: p1 p2
[  100.253209] I/O error, dev mmcblk1, sector 1512 op 0x1:(WRITE) flags 0x4800 phys_seg 63 prio class 2
[  100.253313] Buffer I/O error on dev mmcblk1, logical block 189, lost async page write
[  100.253382] Buffer I/O error on dev mmcblk1, logical block 190, lost async page write
[  100.253440] Buffer I/O error on dev mmcblk1, logical block 191, lost async page write
[  100.253497] Buffer I/O error on dev mmcblk1, logical block 192, lost async page write
[  100.253554] Buffer I/O error on dev mmcblk1, logical block 193, lost async page write
[  100.253610] Buffer I/O error on dev mmcblk1, logical block 194, lost async page write
[  100.253666] Buffer I/O error on dev mmcblk1, logical block 195, lost async page write
[  100.253722] Buffer I/O error on dev mmcblk1, logical block 196, lost async page write
[  100.253779] Buffer I/O error on dev mmcblk1, logical block 197, lost async page write
[  100.253835] Buffer I/O error on dev mmcblk1, logical block 198, lost async page write
[  100.461294] I/O error, dev mmcblk1, sector 3024 op 0x1:(WRITE) flags 0x4800 phys_seg 63 prio class 2
[  100.565894] I/O error, dev mmcblk1, sector 4032 op 0x1:(WRITE) flags 0x4800 phys_seg 63 prio class 2
[  100.598353]  mmcblk1: p1 p2

Any idea on how to fix that?

thank you!

**Reply by DarKOrange75** [reply]
using meson-s4-h96maxw2-50-35-20.dtb fixed the issue.
Bookworm now installed on the eMMC and working like a charm, thank you for your hard work!

**Comment by manu69x** [comment]
Hello bro.
Any progress in wifi driver?
In Corelec now wifi works
udevadm info /sys/bus/sdio/devices/*
give

 /devices/platform/soc/fe088000.sdio/mmc_host/mmc2/mmc2:2ec5/mmc2:2ec5:1
M: mmc2:2ec5:1
R: 1
U: sdio
V: aicwf_sdio
E: DEVPATH=/devices/platform/soc/fe088000.sdio/mmc_host/mmc2/mmc2:2ec5/mmc2:2ec5:1
E: SUBSYSTEM=sdio
E: **DRIVER=aicwf_sdio**
E: OF_NAME=wifi
E: OF_FULLNAME=/soc/sdio@fe088000/wifi@1
E: OF_COMPATIBLE_0=brcm,bcm4329-fmac
E: OF_COMPATIBLE_N=1
E: SDIO_CLASS=07
E: SDIO_ID=5449:0145
E: MODALIAS=sdio:c07v5449d0145
E: USEC_INITIALIZED=18386573
E: ID_SDIO_CLASS_FROM_DATABASE=WLAN interface

**Reply by devmfc** [reply]
I'm not your bro or ho

**Reply by manu69x** [reply]
I'm sorry, I didn't mean to be inappropriate

**Reply by devmfc** [reply]
Ok, glad to hear.

Anyway, I'm only working on WiFi when I really have nothing else to do (and I'm not a WiFi expert, so I can only fix compile bugs or add some porting). That will most certainly not be the case the next coming half year.

It works in CoreElec (which version btw?), great, but CoreElec is based on a old vendor kernel. My images are based on Mainline LTS. So you only have to port that driver from kernel version 5.4 (or 5.15 in case of CE NO) to 6.6 mainline kernel. 
So if you have some time to spare, the needed kernel header files are in the release files. If you get it working I will be more than happy to add it to the images.

**Reply by manu69x** [reply]
First of all thank you for your great work,
thanks to you we were able to install klipper on this cheap TV box, for a school project, in which we improved an old 3D printer.
I'm not a Linux expert, I can at most follow a few guides, I've never compiled a kernel, but I have some time until September until school reopens.
I looked for some step by step guide explaining how to install this wifi driver, but I couldn't find anything suited to my skills.

**Reply by devmfc** [reply]
You're welcome. 

Unfortunately, you can't just install the driver. With one exception: if there is a ready to use DKMS package (which technically still isn't installing). 

Otherwise, most of the cases,  you will have to find the source for a working driver, adapt the source to match the kernel you want to use it for, compile (or cross compile when you are compiling on x86) it, find the right firmware , store it at the right place and then hope for it to work.

I don't know which source CE uses, but again, that is an older and different kernel.
I think you'll have the highest chance with this source: [aic8800](https://github.com/radxa-pkg/aic8800)

To install the needed kernel headers: download the headers package from the release files: [linux-headers-6.6.41-meson64_20240721_arm64.deb](https://github.com/devmfc/debian-on-amlogic/releases/download/v6.6.41/linux-headers-6.6.41-meson64_20240721_arm64.deb). Install via `apt-get install ./linux-headers-6.6.41-meson64_20240721_arm64.deb`.

This is all I can do for you. Good luck!

**Comment by petesimon** [comment]
thanks everyone, especially @devmfc for their hard work. i have a h96 max w2 (2 GB RAM / 16 GB ROM), and i booted ubuntu plucky v6.12.30 / May 25, 2025 from a USB 2.0 stick inserted in the USB 2.0 black port. that works nicely. also booting from the TF/SD slot is possible but not much happens when a slow/cheap card is there. a USB hub, wifi adapter and keyboard are connected to the blue port. Can't get the USB realtek wifi to work either, even though "lsusb" shows the device. I installed the package "xubuntu-desktop" and rebooted and there was a just a blinking cursor __ . Rebooted again via remote SSH and everything went crazy nuts, so that didn't help. I will delete the installation and reinstall as a text headless system only.

Is there a Xorg graphical package that works without re-configuring and tinkering?

![devmfc-login-IMG_20250628_210414](https://github.com/user-attachments/assets/e65162e7-d24c-4ef2-94f5-d893af220f3c)

**Reply by petesimon** [reply]
I got the X lxde graphical desktop working via HDMI. see [this comment in another thread](https://github.com/devmfc/debian-on-amlogic/discussions/2#discussioncomment-13611636). Also got [pihole](https://pi-hole.net/) working by following [this one guide on github](https://github.com/kdrapel/x96mini_linux#readme) and other guides i found in google.

**Comment by Segu1212** [comment]
Excuse me, could someone please send me a copy of the boot0 and boot1 files? The eMMC on this box is shorted out. I left it stored for a while, but when I wanted to learn how to use it, it stopped working. I'm also sharing the ISP connection for this H96MAX W2 box.  ![20260123_190552.jpg](https://github.com/user-attachments/assets/153a04d1-8063-45e1-ac2c-d8aecd9aa554)


![20260123_175739.jpg](https://github.com/user-attachments/assets/f4b64d6d-8b54-4976-b90d-16d61e281523)
